File: validate-transaction/src/test/java/br/com/cardoso/TransactionIntegrationTest.java
package br.com.cardoso;

import br.com.cardoso.dto.InitialTransaction;
import br.com.cardoso.model.TransactionRequestResponseData;
import br.com.cardoso.model.TransactionStatus;
import br.com.cardoso.model.User;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.mockserver.client.MockServerClient;
import org.mockserver.junit.jupiter.MockServerExtension;
import org.mockserver.junit.jupiter.MockServerSettings;
import org.mockserver.matchers.Times;
import org.mockserver.model.HttpTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.kafka.test.context.EmbeddedKafka;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockserver.model.HttpRequest.request;

@ExtendWith(MockServerExtension.class)
@MockServerSettings(ports = {0})
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@EmbeddedKafka(partitions = 1, topics = {"event-hub"})
public class TransactionIntegrationTest {

    @Autowired
    private TestRestTemplate testRestTemplate;
    @Autowired
    private ObjectMapper objectMapper;
    @LocalServerPort
    private int port;
    
    private static final List<TransactionRequestResponseData> transactionRequestResponseDataList = new ArrayList<>();

    @BeforeAll
    static void setup(MockServerClient mockServerClient) {
        String baseUrl = "http://localhost:" + mockServerClient.getPort();
        System.setProperty("transaction.base.url", baseUrl);

        mockServerClient.when(request().withMethod("POST").withPath("/validate/transaction/v1"), Times.exactly(2))
                .respond(HttpTemplate.template(HttpTemplate.TemplateType.VELOCITY,
                        "{'statusCode': 425,'body': 'Falha temporária, tente novamente em alguns instantes.'}"));

        mockServerClient.when(request().withMethod("POST").withPath("/validate/transaction/v1"))
                .respond(HttpTemplate.template(HttpTemplate.TemplateType.VELOCITY,
                        "#set($jsonBody = $json.parse($!request.body)){'statusCode': 200,'body': " +
                                "{'transactionId': '$!uuid', 'value': $jsonBody.value, " +
                                "'user': {'fullname': '$jsonBody.user.fullName', 'document': '$jsonBody.user.document', 'validation': $jsonBody.user.validation}, " +
                                "'transactionStatus': #if($jsonBody.user.validation == -1) 'ERROR' #elseif($jsonBody.user.validation == 0) 'DENIED' #else 'AUTHORIZED' #end}}"));
    }

    @AfterAll
    static void shouldValidateAllTransactionStatusWhenListReceivedKafkaMessages() {
        //given
        await().atMost(10, SECONDS).until(() -> transactionRequestResponseDataList.size() == 4);
        List<String> responseList = transactionRequestResponseDataList.stream().map(TransactionRequestResponseData::responseBody).toList();

        //then
        assertTrue(responseList.stream().anyMatch(response -> response.contains(TransactionStatus.ERROR.name())),
                "Esperado que ao menos uma resposta tenha o status de 'ERROR'");
        assertTrue(responseList.stream().anyMatch(response -> response.contains(TransactionStatus.AUTHORIZED.name())),
                "Esperado que ao menos uma resposta tenha o status de 'AUTHORIZED'");
        assertTrue(responseList.stream().anyMatch(response -> response.contains(TransactionStatus.DENIED.name())),
                "Esperado que ao menos uma resposta tenha o status de 'DENIED'");
        assertFalse(responseList.stream().anyMatch(response -> response.contains(TransactionStatus.CREATED.name())),
                "Nenhuma resposta deve ter o status 'CREATED'");
    }

    @ParameterizedTest
    @EnumSource(TransactionStatus.class)
    void shouldReturnTransactionStatusWhenTransactionIsSent(TransactionStatus expectedStatus) throws JsonProcessingException {
        //given
        BigDecimal value = new BigDecimal(1000);
        int validation = switch (expectedStatus) {
            case DENIED -> 0;
            case AUTHORIZED -> 1;
            default -> -1;
        };

        User user = new User("John Doe", "123456789", validation);
        InitialTransaction initialTransaction = new InitialTransaction(value, user);

        String requestBody = objectMapper.writeValueAsString(initialTransaction);
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        HttpEntity<String> entity = new HttpEntity<>(requestBody, headers);

        //then
        ResponseEntity<TransactionStatus> response = testRestTemplate.postForEntity("http://localhost:" + port + "/transaction/v1", entity, TransactionStatus.class);

        //No cenário integrado, a transação nunca deve retornar com o status CREATED
        if (expectedStatus.equals(TransactionStatus.CREATED)) {
            assertEquals(TransactionStatus.ERROR, response.getBody());
        } else {
            assertEquals(expectedStatus, response.getBody());
        }
    }

    @KafkaListener(groupId = "transactionGroup", topics = "event-hub")
    public void listen(ConsumerRecord<String, String> record) throws JsonProcessingException {
        transactionRequestResponseDataList.add(objectMapper.readValue(record.value(), TransactionRequestResponseData.class));
    }
}

File: transaction-recorder/src/test/kotlin/br/com/cardoso/TransactionIntegrationTest.kt
package br.com.cardoso

import br.com.cardoso.dto.Transaction
import br.com.cardoso.dto.TransactionRevision
import br.com.cardoso.listener.KafkaTransactionListener
import br.com.cardoso.model.TransactionRequest
import br.com.cardoso.model.TransactionRequestResponseData
import br.com.cardoso.model.TransactionResponse
import br.com.cardoso.model.User
import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.test.web.server.LocalServerPort
import org.springframework.core.ParameterizedTypeReference
import org.springframework.core.env.Environment
import org.springframework.data.history.RevisionMetadata.RevisionType
import org.springframework.http.HttpMethod
import org.springframework.http.ResponseEntity
import org.springframework.kafka.test.context.EmbeddedKafka
import org.springframework.test.context.ActiveProfiles
import java.math.BigDecimal

@SpringBootTest(
    webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
    properties = [
        "spring.datasource.url=jdbc:tc:mysql:8.4.0:///db",
        "spring.datasource.driver-class-name=org.testcontainers.jdbc.ContainerDatabaseDriver"
    ]
)
@EmbeddedKafka(partitions = 1, topics = ["event-hub"])
@ActiveProfiles("test")
class TransactionIntegrationTest {

    @Autowired
    lateinit var objectMapper: ObjectMapper
    @Autowired
    lateinit var kafkaTransactionListener: KafkaTransactionListener
    @Autowired
    lateinit var environment: Environment
    @Autowired
    lateinit var testRestTemplate: TestRestTemplate
    @LocalServerPort
    private val port = 0

    @Test
    fun `should process transaction and validate audit`() {
        //given
        val user = User(name = "John Doe", document = "123456789", validation = 1)
        val (transactionRequest, transactionResponse, transactionRequestResponseData) = generateUserResponseTransactionRequestData(user)
        sendKafkaMessage(transactionRequestResponseData)

        //when
        val allTransactionsResponse: ResponseEntity<List<Transaction>> = testRestTemplate.exchange(
            "http://localhost:$port/v1/transactions",
            HttpMethod.GET,
            null,
            object : ParameterizedTypeReference<List<Transaction>>() {}
        )

        //then
        assertEquals(200, allTransactionsResponse.statusCode.value())
        assertNotNull(allTransactionsResponse.body)
        assertEquals(1, allTransactionsResponse.body!!.size)
        assertEquals(0, transactionResponse.value.compareTo(allTransactionsResponse.body!![0].value))
        assertEquals("123XXXXXX", allTransactionsResponse.body!![0].userDocument)
        assertEquals(transactionResponse.transactionId, allTransactionsResponse.body!![0].transactionId)
        assertEquals(transactionResponse.transactionStatus, allTransactionsResponse.body!![0].transactionStatus)
        assertEquals(user.name, allTransactionsResponse.body!![0].userName)
        assertEquals(user.document, allTransactionsResponse.body!![0].userDocument)
        assertEquals(1, user.validation)
        assertEquals("https://request.uri", transactionRequestResponseData.requestUri)
        assertEquals("POST", transactionRequestResponseData.requestMethod)
        assertEquals(200, transactionRequestResponseData.responseStatus)

        //given
        val savedTransactionId = allTransactionsResponse.body!![0].id

        //when
        val deleteResponse: ResponseEntity<Void> = testRestTemplate.exchange(
            "http://localhost:$port/v1/transactions/$savedTransactionId",
            HttpMethod.DELETE,
            null,
            Void::class.java
        )
        val allTransactionsResponseAfterDelete: ResponseEntity<List<Transaction>> = testRestTemplate.exchange(
            "http://localhost:$port/v1/transactions",
            HttpMethod.GET,
            null,
            object : ParameterizedTypeReference<List<Transaction>>() {}
        )

        //then
        assertEquals(204, deleteResponse.statusCode.value())
        assertTrue(allTransactionsResponseAfterDelete.body!!.isEmpty())

        //when
        val allRevisionsResponseAfterDelete: ResponseEntity<List<TransactionRevision>> = testRestTemplate.exchange(
            "http://localhost:$port/v1/transactions/$savedTransactionId",
            HttpMethod.GET,
            null,
            object : ParameterizedTypeReference<List<TransactionRevision>>() {}
        )

        //then
        assertEquals(200, allRevisionsResponseAfterDelete.statusCode.value())
        assertNotNull(allRevisionsResponseAfterDelete.body)
        assertEquals(2, allRevisionsResponseAfterDelete.body!!.size)

        assertEquals(0, transactionResponse.value.compareTo(allRevisionsResponseAfterDelete.body!![0].value))
        assertEquals("123XXXXXX", allRevisionsResponseAfterDelete.body!![0].userDocument)
        assertEquals(transactionResponse.transactionId, allRevisionsResponseAfterDelete.body!![0].transactionId)
        assertEquals(transactionResponse.transactionStatus, allRevisionsResponseAfterDelete.body!![0].transactionStatus)
        assertNotEquals(transactionRequest.transactionStatus, allRevisionsResponseAfterDelete.body!![0].transactionStatus)
        assertEquals(user.name, allRevisionsResponseAfterDelete.body!![0].userName)
        assertEquals(0, allRevisionsResponseAfterDelete.body!![0].version)
        assertEquals(RevisionType.INSERT, allRevisionsResponseAfterDelete.body!![0].revisionType)

        assertNull(allRevisionsResponseAfterDelete.body!![1].value)
        assertNull(allRevisionsResponseAfterDelete.body!![1].userDocument)
        assertNull(allRevisionsResponseAfterDelete.body!![1].transactionId)
        assertNull(allRevisionsResponseAfterDelete.body!![1].transactionStatus)
        assertNull(allRevisionsResponseAfterDelete.body!![1].userName)
        assertEquals(0, allRevisionsResponseAfterDelete.body!![1].version)
        assertEquals(RevisionType.DELETE, allRevisionsResponseAfterDelete.body!![1].revisionType)
    }

    private fun sendKafkaMessage(transactionRequestResponseData: TransactionRequestResponseData) {
        //Não enviado via Kafka template para não ter a obrigatoriedade da configuração do producer e aguardar pelo recebimento da mensagem
        val consumerRecordTransaction: ConsumerRecord<String, TransactionRequestResponseData> =
            ConsumerRecord("event-hub", 0, 0, "key", transactionRequestResponseData)
        kafkaTransactionListener.listen(consumerRecordTransaction)
    }

    private fun generateUserResponseTransactionRequestData(user: User): Triple<TransactionRequest, TransactionResponse, TransactionRequestResponseData> {
        val transactionRequest = TransactionRequest(
            transactionStatus = "CREATED",
            value = BigDecimal(1000.00),
            user = user
        )
        val transactionResponse = TransactionResponse(
            transactionId = "123",
            transactionStatus = "AUTHORIZED",
            value = BigDecimal(1000.00),
            user = user
        )
        val transactionRequestResponseData = TransactionRequestResponseData(
            requestUri = "https://request.uri",
            requestMethod = "POST",
            requestBody = objectMapper.writeValueAsString(transactionRequest),
            responseBody = objectMapper.writeValueAsString(transactionResponse),
            responseStatus = 200
        )
        return Triple(transactionRequest, transactionResponse, transactionRequestResponseData)
    }
}